c/certificateService.c

```diff
@@ -0,0 +1,187 @@
+    setResponseStatus(response, 401, "Unauthorized");
+  } else {
+    setResponseStatus(response, 500, "Internal server error");
+  }
+}
+
+void handleInvalidMethod(HttpResponse *response) {
+    jsonPrinter *p = respondWithJsonPrinter(response);
+      
+    setResponseStatus(response, 405, "Method Not Allowed");
+    setDefaultJSONRESTHeaders(response);
+    addStringHeader(response, "Allow", "GET");
+    writeHeader(response);
+    
+    jsonStart(p);
+    {
+      jsonAddString(p, "error", "Only POST requests are supported");
+    }
+    jsonEnd(p);
+}
+
+static int serveMappingService(HttpService *service, HttpResponse *response)
+{
+  HttpRequest *request = response->request;
+  char *routeFragment = stringListPrint(request->parsedFile, 1, 1000, "/", 0);
+  char *route = stringConcatenate(response->slh, "/", routeFragment);
+    
+  if (!strcmp(request->method, methodPOST))
+  {
+    char *inPtr = request->contentBody;
+    if(request->contentLength >= 4096) {
+      setResponseStatus(response, 400, "Bad request");
+      finishResponse(response);
+    }
+    
+    R_datalib_parm_list_64 example;
+    memset(&example, 0, sizeof(R_datalib_parm_list_64));
+
+    example.certificate_len = request->contentLength;
+    memcpy(example.certificate, inPtr, request->contentLength);
+
+    example.function_code = 0x0006;
+    int rc; 
+
+    rc = IRRSIM00(
+        &example.workarea, // WORKAREA 
+        &example.saf_rc_ALET  , // ALET 
+        &example.return_code, 
+        &example.racf_rc_ALET, 
+        &example.RACF_return_code,
+        &example.racf_rsn_ALET,
+        &example.RACF_reason_code,
+        &example.fc_ALET,
+        &example.function_code,
+        &example.option_word,
+        &example.RACF_userid_len,
+        &example.certificate_len,
+        &example.application_id_len,
+        &example.distinguished_name_len,
+        &example.registry_name_len
+    );
+      
+    jsonPrinter *p = respondWithJsonPrinter(response);
+    
+    setValidResponseCode(response, rc, example.return_code, example.RACF_return_code, example.RACF_reason_code);
+    setDefaultJSONRESTHeaders(response);
+    writeHeader(response);
+    
+    jsonStart(p);
+    {
+      jsonAddString(p, "userid", example.RACF_userid);
+      jsonAddInt(p, "rc", rc);
+      jsonAddInt(p, "saf_rc", example.return_code);
+      jsonAddInt(p, "racf_rc", example.RACF_return_code);
+      jsonAddInt(p, "reason_code", example.RACF_reason_code);
+    }
+    jsonEnd(p);
+  }
+  else 
+  {
+     handleInvalidMethod(response);
+  }
+  
+  finishResponse(response);
+    
+  return 0;
+}
+
+void installCertificateService(HttpServer *server)
+{
+  HttpService *httpService = makeGeneratedService("CertificateService", "/certificate/**");
+  httpService->authType = SERVICE_AUTH_NONE;
+  httpService->serviceFunction = serveMappingService;
+  httpService->runInSubtask = TRUE;
+  httpService->doImpersonation = FALSE;
```

Comment:
Ultimately I don't know. 
What we need is to be able to call the R_usermap macro. By we I mean the API Mediation Layer service, as there is no authentication at the moment, who will be the user under which this will be called?

Reply:
In the future (and perhaps in this case) it is worth submitting a text-format functional spec for review in advance of related implementation source code.  Implementing unauthenticated, publicly-accessible wrappers around authorized SAF services is not a pattern we should *ever* adopt without strict scrutiny and a thorough vetting of the related cost/benefit analysis from the customer's perspective.